[SwiftUI] Add plumbing and test infrastructure to support more accurate scroll geometry reporting

pxlcoder opened the pull request:
#### 7cfc79c2b3d82989bd001597f61053ecb65ff19c
<pre>
[SwiftUI] Add plumbing and test infrastructure to support more accurate scroll geometry reporting
<a href="https://bugs.webkit.org/show_bug.cgi?id=294694">https://bugs.webkit.org/show_bug.cgi?id=294694</a>
<a href="https://rdar.apple.com/153764316">rdar://153764316</a>

Reviewed by Richard Robinson, Wenson Hsieh, and Megan Gardner.

`webViewOnScrollGeometryChange` currently does not always report an accurate
content size.  The reported size should be a &quot;fitting size&quot; given the current
web view size. However, at this time the reported size is always the `FrameView`&apos;s
content size, which is often larger.

This patch works towards fixing the reported size by introducing plumbing to
only compute a content size for scroll geometry if needed, and reporting that
size back to SwiftUI. Additionally, testing capability for scroll geometry is
added.

A subsequent patch will fill out the implementation of
`LocalFrameView::updateScrollGeometryContentSize` to fix the described issue.
That patch will also introduce even more test cases, including ones that are
currently broken.

No behavior change.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::needsScrollGeometryUpdates const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
(WebCore::LocalFrameView::updateScrollGeometryContentSize):

Recompute the content size for SwiftUI after every layout, but only if
scroll geometry updates are needed. This ensures expensive (to be added)
work is avoided when unnecessary.

For now, simply return `contentsSize()`, preserving the currently
implemented behavior.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:

Add a new field to `RemoteLayerTreeTransaction` so that content size changes
are reported in-sync with scroll position changes.

(WebKit::RemoteLayerTreeTransaction::scrollGeometryContentSize const):
(WebKit::RemoteLayerTreeTransaction::setScrollGeometryContentSize):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setNeedsScrollGeometryUpdates:]):

Add a hook for SwiftUI to let WebKit know scroll geometry updates are needed.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
* Source/WebKit/UIProcess/Cocoa/WebPageWebView.swift:
(WebPageWebView.setNeedsScrollGeometryUpdates(_:)):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setNeedsScrollGeometryUpdates):
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::needsScrollGeometryUpdates const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::willCommitLayerTree):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/_WebKit_SwiftUI/Implementation/CocoaWebViewAdapter.swift:
(CocoaWebViewAdapter.onScrollGeometryChange):

Only perform scroll geometry content size computation if the view modifier is
in use.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKScrollGeometryTests.mm: Added.
(-[TestScrollGeometryDelegate _webView:geometryDidChange:]):
(-[TestScrollGeometryDelegate currentContentSize]):
(runContentSizeTest):

Test both quirks and standards mode.

(TEST(WKScrollGeometry, ContentSizeTallerThanWebView)):
(TEST(WKScrollGeometry, NoScrollGeometryUpdates)):

Canonical link: <a href="https://commits.webkit.org/296531@main">https://commits.webkit.org/296531@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3d85a62180b2dc8a56089547b2e1fdc38f89aec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108420 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113629 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58833 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110383 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28770 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36631 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82333 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111368 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22817 "Found 3 new test failures: fast/filter-image/clipped-filter.html http/tests/geolocation/geolocation-get-current-position-does-not-leak.https.html http/tests/security/cached-svg-image-with-css-cross-domain.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97658 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62769 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22233 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15796 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58355 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92186 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116750 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35474 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26134 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91360 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35847 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93933 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91161 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36050 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13813 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31214 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17568 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35377 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40914 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35088 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38438 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36771 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -1295,6 +1297,18 @@ void LocalFrameView::didLayout(SingleThreadWeakPtr<RenderElement> layoutRoot, bo
+    if (!m_frame->isMainFrame())
+        return;
+
+    RefPtr page = m_frame->page();
+    if (!page || !page->chrome().client().needsScrollGeometryUpdates())
+        return;
+
+    m_scrollGeometryContentSize = contentsSize();
```
stylistic nit:
```suggestion
    if (RefPtr page = m_frame->page(); page && page->chrome().client().needsScrollGeometryUpdates())
        m_scrollGeometryContentSize = contentsSize();
```

pxlcoder replied:
I won't be making that change, because the next patch will remove line 1309 and replace it with the actual algorithm 😅

pxlcoder committed:
[SwiftUI] Add plumbing and test infrastructure to support more accurate scroll geometry reporting
https://bugs.webkit.org/show_bug.cgi?id=294694
rdar://153764316

Reviewed by Richard Robinson, Wenson Hsieh, and Megan Gardner.

`webViewOnScrollGeometryChange` currently does not always report an accurate
content size.  The reported size should be a "fitting size" given the current
web view size. However, at this time the reported size is always the `FrameView`'s
content size, which is often larger.

This patch works towards fixing the reported size by introducing plumbing to
only compute a content size for scroll geometry if needed, and reporting that
size back to SwiftUI. Additionally, testing capability for scroll geometry is
added.

A subsequent patch will fill out the implementation of
`LocalFrameView::updateScrollGeometryContentSize` to fix the described issue.
That patch will also introduce even more test cases, including ones that are
currently broken.

No behavior change.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::needsScrollGeometryUpdates const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::didLayout):
(WebCore::LocalFrameView::updateScrollGeometryContentSize):

Recompute the content size for SwiftUI after every layout, but only if
scroll geometry updates are needed. This ensures expensive (to be added)
work is avoided when unnecessary.

For now, simply return `contentsSize()`, preserving the currently
implemented behavior.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:

Add a new field to `RemoteLayerTreeTransaction` so that content size changes
are reported in-sync with scroll position changes.

(WebKit::RemoteLayerTreeTransaction::scrollGeometryContentSize const):
(WebKit::RemoteLayerTreeTransaction::setScrollGeometryContentSize):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setNeedsScrollGeometryUpdates:]):

Add a hook for SwiftUI to let WebKit know scroll geometry updates are needed.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
* Source/WebKit/UIProcess/Cocoa/WebPageWebView.swift:
(WebPageWebView.setNeedsScrollGeometryUpdates(_:)):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setNeedsScrollGeometryUpdates):
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::needsScrollGeometryUpdates const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
(WebKit::WebPage::willCommitLayerTree):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/_WebKit_SwiftUI/Implementation/CocoaWebViewAdapter.swift:
(CocoaWebViewAdapter.onScrollGeometryChange):

Only perform scroll geometry content size computation if the view modifier is
in use.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKScrollGeometryTests.mm: Added.
(-[TestScrollGeometryDelegate _webView:geometryDidChange:]):
(-[TestScrollGeometryDelegate currentContentSize]):
(runContentSizeTest):

Test both quirks and standards mode.

(TEST(WKScrollGeometry, ContentSizeTallerThanWebView)):
(TEST(WKScrollGeometry, NoScrollGeometryUpdates)):

Canonical link: https://commits.webkit.org/296531@main
